### PR TITLE
Add button to duplicate job definition

### PR DESCRIPTION
**Pull request by takonomura**
Add button for duplicate job definition to Job Definition Details page.
This allows us to duplicate the definition for similarly jobs.
On press the button, it open New Job Definition page with pre-filled the definition.

Before:
before
After:
after

### CODE REVIEW

**Comment by ganmacs** [comment]
@cookpad/dev-infra review please

File: app/models/kuroko2/job_definition.rb [hunk 1]
```diff
@@ -81,7 +81,7 @@ def proceed_multi_instance?
   end
 
   def text_tags
-    tags.pluck(:name).join(',')
+    tags.map(&:name).join(',')
```
Comment (eagletmt) [inline app/models/kuroko2/job_definition.rb]: Why this change needed?

Comment (takonomura) [reply app/models/kuroko2/job_definition.rb]: Without this change, `text_tags` of duplicated definition always returns empty.

```ruby
definition = Kuroko2::JobDefinition.first
duplicated_definition = definition.dup
duplicated_definition.tags = definition.tags.dup

definition.tags.pluck(:name) #=> ["HelloWorld"]
definition.tags.map(&:name) #=> ["HelloWorld"]
duplicated_definition.tags.pluck(:name) #=> []
duplicated_definition.tags.map(&:name) #=> ["HelloWorld"]
```

File: spec/controllers/job_definitions_controller_spec.rb [hunk 2]
```diff
@@ -38,6 +38,64 @@
...
     end
+
+    context 'with dup_from params' do
+      subject! { get :new, params: { dup_from: definition.id } }
+      it do
+        expect(response).to have_http_status(:ok)
+        expect(response).to render_template('new')
+
+        expect(assigns(:definition)).to be_new_record
+        expect(assigns(:definition).name).to eq definition.name
```
Comment (ganmacs) [inline spec/controllers/job_definitions_controller_spec.rb]: Could you write expect expressions about `tags` and  `admins` ?

File: spec/controllers/job_definitions_controller_spec.rb [hunk 3]
```diff
@@ -38,6 +38,64 @@
...
+        it do
+          expect(response).to have_http_status(:ok)
+          expect(response).to render_template('new')
+
+          expect(assigns(:definition)).to be_new_record
+          expect(assigns(:definition).name).to eq definition.name
+          expect(assigns(:definition).admins).to match_array [admin_user, controller.current_user]
+        end
+
+        context 'with other admins' do
```
Comment (ganmacs) [inline spec/controllers/job_definitions_controller_spec.rb]: `'with only anther admin user'` ?

File: spec/controllers/job_definitions_controller_spec.rb [hunk 4]
```diff
@@ -38,6 +38,64 @@
...
+      subject! { get :new, params: { dup_from: definition.id } }
+      it do
+        expect(response).to have_http_status(:ok)
+        expect(response).to render_template('new')
+
+        expect(assigns(:definition)).to be_new_record
+        expect(assigns(:definition).name).to eq definition.name
+      end
+
+      context 'with admins' do
```
Comment (ganmacs) [inline spec/controllers/job_definitions_controller_spec.rb]: `'with current_user and another admin user' `

File: spec/controllers/job_definitions_controller_spec.rb (hunk 3 above)
Comment (takonomura) [reply spec/controllers/job_definitions_controller_spec.rb]: Fixed

File: spec/controllers/job_definitions_controller_spec.rb [hunk 5]
```diff
@@ -38,6 +38,67 @@
...
+        end
+      end
+
+      context 'with invalid id' do
+        subject! { get :new, params: { dup_from: 'invalid' } }
+        it do
+          expect(response).to have_http_status(:ok)
+          expect(response).to render_template('new')
+
+          expect(assigns(:definition)).to be_new_record
```
Comment (eisuke) [inline spec/controllers/job_definitions_controller_spec.rb]: It seems better to check what a field is blank.

File: spec/controllers/job_definitions_controller_spec.rb [hunk 6]
```diff
@@ -38,6 +38,67 @@
...
+
+    context 'with dup_from params' do
+      subject! { get :new, params: { dup_from: definition.id } }
+      it do
+        expect(response).to have_http_status(:ok)
+        expect(response).to render_template('new')
+
+        expect(assigns(:definition)).to be_new_record
+        expect(assigns(:definition).name).to eq definition.name
+        definition.admins << controller.current_user unless definition.admins.include?(controller.current_user)
```
Comment (eisuke) [inline spec/controllers/job_definitions_controller_spec.rb]: It seems to better for do not change `definition` instance in `it` block.
How about changing it to the below code?

```ruby
expect(assigns(:definition).admins).to eq(definition.admins + controller.current_user)
```

File: app/models/kuroko2/job_definition.rb (hunk 1 above)
Comment (ganmacs) [reply app/models/kuroko2/job_definition.rb]: `text_tags` method is called at [here](https://github.com/cookpad/kuroko2/blob/52c077269f7613546f88152e00e02c3862e50098/app/views/kuroko2/job_definitions/_form.html.slim#L85).
In [new page](https://github.com/cookpad/kuroko2/blob/52c077269f7613546f88152e00e02c3862e50098/app/views/kuroko2/job_definitions/new.html.slim),  `text_tags` would return  empty array without this change because the job_definition having this `tags` does not exists in DB (so `#pluck` returns empty array).   

Is it right? @takonomura

Comment (takonomura) [reply app/models/kuroko2/job_definition.rb]: Yes

File: spec/controllers/job_definitions_controller_spec.rb [hunk 7]
```diff
@@ -38,6 +38,68 @@
...
+
+    context 'with dup_from params' do
+      subject! { get :new, params: { dup_from: definition.id } }
+      it do
+        expect(response).to have_http_status(:ok)
+        expect(response).to render_template('new')
+
+        expect(assigns(:definition)).to be_new_record
+        expect(assigns(:definition).name).to eq definition.name
+        expect(assigns(:definition).admins).to eq definition.admins + [controller.current_user]
```
Comment (ganmacs) [inline spec/controllers/job_definitions_controller_spec.rb]: [nis]
[definition.admins, controller.current_user] seems to be better.

File: spec/controllers/job_definitions_controller_spec.rb [hunk 8]
```diff
@@ -38,6 +38,68 @@
...
+        expect(assigns(:definition).tags).to eq definition.tags
+      end
+
+      context 'with current_user and aother admin user' do
```
Comment (ganmacs) [inline spec/controllers/job_definitions_controller_spec.rb]: [typo] another?

Comment (takonomura) [reply spec/controllers/job_definitions_controller_spec.rb]: Thanks, I fixed it in https://github.com/cookpad/kuroko2/pull/92/commits/33ff102f4903dff4accb77200681ae8bdd70de3f.

File: spec/controllers/job_definitions_controller_spec.rb (hunk 7 above)
Comment (takonomura) [reply spec/controllers/job_definitions_controller_spec.rb]: `expect(assigns(:definition).admins).to eq [definition.admins, controller.current_user]` is not working.

Comment (ganmacs) [reply spec/controllers/job_definitions_controller_spec.rb]: oh..sorry, I misunderstood.  `definition.admins` is array.
